### PR TITLE
Units

### PR DESCRIPTION
**Pull request by gertjanvanzwieten**

### CODE REVIEW

File: nutils/expression_v2.py [hunk 1]
```diff
@@ -792,20 +791,24 @@ def jump(self, array: function.Array) -> function.Array:
...
-    assert denominator.ndim == 0
-    return function.divide(numerator, function._append_axes(denominator, numerator.shape))
+    assert numpy.ndim(denominator) == 0
+    return numpy.true_divide(numerator, numpy.broadcast_to(denominator, numpy.shape(denominator) + numpy.shape(numerator)))
 
   def power(self, base: function.Array, exponent: function.Array) -> function.Array:
-    assert exponent.ndim == 0
-    return function.power(base, function._append_axes(exponent, base.shape))
+    assert numpy.ndim(exponent) == 0
+    return numpy.power(base, numpy.broadcast_to(exponent, numpy.shape(exponent) + numpy.shape(base)))
```
Comment (joostvanzwieten) [inline nutils/expression_v2.py:814]: `exponent` has no shape, so you can simply use `numpy.broadcast_to(exponent, numpy.shape(base))`. Or let `numpy.power` do the magic.

File: nutils/expression_v2.py [hunk 2]
```diff
@@ -792,20 +791,24 @@ def jump(self, array: function.Array) -> function.Array:
...
     for arg in args[1:]:
-      result = function.multiply(function._append_axes(result, arg.shape), function._prepend_axes(arg, result.shape))
+      result = numpy.multiply(self.append_axes(result, numpy.shape(arg)), arg)
     return result
 
   def divide(self, numerator: function.Array, denominator: function.Array) -> function.Array:
-    assert denominator.ndim == 0
-    return function.divide(numerator, function._append_axes(denominator, numerator.shape))
+    assert numpy.ndim(denominator) == 0
+    return numpy.true_divide(numerator, numpy.broadcast_to(denominator, numpy.shape(denominator) + numpy.shape(numerator)))
```
Comment (joostvanzwieten) [inline nutils/expression_v2.py:810]: Same as `power`.

**Review by joostvanzwieten — APPROVED** [review]

File: nutils/expression_v2.py (hunk 1 above)
Comment (gertjanvanzwieten) [reply nutils/expression_v2.py]: fixed

File: nutils/expression_v2.py (hunk 2 above)
Comment (gertjanvanzwieten) [reply nutils/expression_v2.py]: fixed

File: .github/workflows/test.yaml [hunk 3]
```diff
@@ -134,7 +134,7 @@ jobs:
         id: install
         run: |
           python -um pip install --upgrade wheel
-          python -um pip install --upgrade treelog stringly matplotlib scipy pillow numpy
+          python -um pip install --upgrade treelog stringly matplotlib scipy pillow numpy git+https://github.com/evalf/nutils-SI.git@initialize
```
Comment (joostvanzwieten) [inline .github/workflows/test.yaml:137]: Can you add `nutils-SI` to the container image as well? This seems to work:

```diff
diff --git a/devtools/container/build.py b/devtools/container/build.py
index 04a9c62f..cdd4e323 100644
--- a/devtools/container/build.py
+++ b/devtools/container/build.py
@@ -57,9 +57,10 @@ with ExitStack() as stack:
   if not examples.exists():
     raise SystemExit(f'cannot find example dir: {examples}')

-  container = stack.enter_context(Container.new_from(base, network='none', mounts=[Mount(src=dist, dst='/mnt')]))
+  container = stack.enter_context(Container.new_from(base, mounts=[Mount(src=dist, dst='/mnt')]))

   container.run('pip', 'install', '--no-cache-dir', '--no-index', '--find-links', 'file:///mnt/', 'nutils', env=dict(PYTHONHASHSEED='0'))
+  container.run('pip', 'install', '--no-cache-dir', 'https://github.com/evalf/nutils-SI/archive/initialize.tar.gz', env=dict(PYTHONHASHSEED='0'))
   container.add_label('org.opencontainers.image.url', 'https://github.com/evalf/nutils')
   container.add_label('org.opencontainers.image.source', 'https://github.com/evalf/nutils')
   container.add_label('org.opencontainers.image.authors', 'Evalf')
```

**Review by joostvanzwieten — APPROVED** [review]

File: nutils/function.py [hunk 4]
```diff
@@ -1079,6 +1079,18 @@ def asarray(__arg: IntoArray) -> Array:
 
   return Array.cast(__arg)
 
+@implements(numpy.shape)
+def shape(__arg: IntoArray) -> Array:
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/function.py:1089]: fixed

File: nutils/function.py [hunk 5]
```diff
@@ -1079,6 +1079,18 @@ def asarray(__arg: IntoArray) -> Array:
...
+    return asarray(__arg).shape
+
+@implements(numpy.ndim)
+def ndim(__arg: IntoArray) -> Array:
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/function.py:1093]: fixed

File: nutils/function.py [hunk 6]
```diff
@@ -1079,6 +1079,18 @@ def asarray(__arg: IntoArray) -> Array:
...
+    return asarray(__arg).ndim
+
+@implements(numpy.size)
+def size(__arg: IntoArray) -> Array:
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/function.py:1097]: fixed

File: nutils/function.py [hunk 7]
```diff
@@ -1249,7 +1252,10 @@ def divide(__dividend: IntoArray, __divisor: IntoArray) -> Array:
   :class:`Array`
   '''
 
-  return multiply(__dividend, reciprocal(__divisor))
+  dividend, divisor = typecast_arrays(__dividend, __divisor, min_dtype=float)
+  return reciprocal(divisor) if _is_unit_scalar(__dividend) \
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/function.py:1262]: fixed

File: .github/workflows/test.yaml (hunk 3 above)
Comment (gertjanvanzwieten) [reply .github/workflows/test.yaml:137]: fixed

**Pull request merged**